support/src/main/java/com/yubico/yubikit/support/DeviceUtil.java

```diff
@@ -0,0 +1,528 @@
+     * @throws InvalidParameterException in case of connection is not {@link SmartCardConnection}, {@link OtpConnection} or
+     * {@link FidoConnection}
+     */
+    public static DeviceInfo readInfo(YubiKeyConnection connection, @Nullable YubiKeyUsbProductId pid)
+            throws IOException, InvalidParameterException {
+
+        YubiKeyType keyType = null;
+        int interfaces = 0;
+
+        if (pid != null) {
+            keyType = pid.type;
+            interfaces = pid.usbInterfaces;
+        } else if (!(connection instanceof SmartCardConnection) ||
+                ((SmartCardConnection) connection).getTransport() == Transport.USB) {
+            throw new InvalidParameterException("pid missing for usb connection");
+        }
+
+        DeviceInfo info;
+        if (connection instanceof SmartCardConnection) {
+            info = readInfoCcid((SmartCardConnection) connection, interfaces);
+        } else if (connection instanceof OtpConnection) {
+            info = readInfoOtp((OtpConnection) connection, keyType, interfaces);
+        } else if (connection instanceof FidoConnection) {
+            info = readInfoFido((FidoConnection) connection, keyType);
+        } else {
+            throw new InvalidParameterException("Invalid connection type");
+        }
+
+        Logger.d("Read info " + info);
+
+        final DeviceConfig config = info.getConfig();
+        final Version version = info.getVersion();
+        final FormFactor formFactor = info.getFormFactor();
+
+        @SuppressWarnings("WrapperTypeMayBePrimitive")
+        Integer supportedUsbCapabilities = info.getSupportedCapabilities(Transport.USB);
+        Integer supportedNfcCapabilities = info.getSupportedCapabilities(Transport.NFC);
+
+        Integer enabledUsbCapabilities = config.getEnabledCapabilities(Transport.USB);
+        Integer enabledNfcCapabilities = config.getEnabledCapabilities(Transport.NFC);
+
+        // Set usbEnabled if missing (pre YubiKey 5)
+        if (info.hasTransport(Transport.USB) && enabledUsbCapabilities == null) {
+
+            int usbEnabled = supportedUsbCapabilities;
+            if (usbEnabled == (OTP.bit | U2F.bit | UsbInterface.CCID)) {
+                // YubiKey Edge, hide unusable CCID interface from supported
+                supportedUsbCapabilities = OTP.bit | U2F.bit;
+            }
+
+            if ((interfaces & UsbInterface.OTP) == 0) {
+                usbEnabled &= ~OTP.bit;
+            }
+
+            if ((interfaces & UsbInterface.FIDO) == 0) {
+                usbEnabled &= ~(U2F.bit | FIDO2.bit);
+            }
+
+            if ((interfaces & UsbInterface.CCID) == 0) {
+                usbEnabled &= ~(UsbInterface.CCID | OATH.bit | OPENPGP.bit | PIV.bit);
+            }
+
+            enabledUsbCapabilities = usbEnabled;
+        }
+
+        boolean isSky = info.isSky();
+        if (keyType == YubiKeyType.SKY) {
+            isSky = true;
+        }
+
+        boolean isFips = info.isFips();
```

Comment:
Could be shortened to (and be made final) `isFips= info.isFips() || (version.major == 4 && version.minor == 4);`